src/main/java/com/treasuredata/bigdam/log/MetricMonitor.java

```diff
@@ -0,0 +1,170 @@
+package com.treasuredata.bigdam.log;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+public class MetricMonitor
+{
+    @FunctionalInterface
+    public interface MetricProducer
+    {
+        public Map<String, Object> produce();
+    }
+
+    @FunctionalInterface
+    public interface ComplexMetricProducer
+    {
+        public List<ComplexMetric> produce();
+    }
+
+    private final Log logger;
+
+    private final String tagPrefixMetric;
+    private final String tagPrefixRawMetric;
+    private final String metricFieldName;
+
+    private final List<MetricProducer> metricProducers;
+    private final List<MetricProducer> rawMetricProducers;
+    private final List<ComplexMetricProducer> metricProducersComplex;
+    private final List<ComplexMetricProducer> rawMetricProducersComplex;
+
+    private final int metricIntervalSeconds;
+
+    private final AtomicBoolean running;
+    private Thread monitorThread;
+    private long sleepIntervalMilliSeconds;
+
+    private static final long SLEEP_INTERVAL_MSEC = 800L;
+    private static final long MONITOR_THREAD_STOP_TIMEOUT = 2000L; // sleep interva * 2 + alpha
+
+    public MetricMonitor(
+            final Log logger,
+            final String tagPrefixMetric,
+            final String tagPrefixRawMetric,
+            final String metricFieldName,
+            final int metricIntervalSeconds
+    )
+    {
+        this.logger = logger;
+
+        this.tagPrefixMetric = tagPrefixMetric;
+        this.tagPrefixRawMetric = tagPrefixRawMetric;
+        this.metricFieldName = metricFieldName;
+
+        this.metricIntervalSeconds = metricIntervalSeconds;
+        this.sleepIntervalMilliSeconds = SLEEP_INTERVAL_MSEC;
+
+        this.metricProducers = new ArrayList<>();
+        this.rawMetricProducers = new ArrayList<>();
+        this.metricProducersComplex = new ArrayList<>();
+        this.rawMetricProducersComplex = new ArrayList<>();
+
+        this.running = new AtomicBoolean(false);
+    }
+
+    public void addMetricProducer(final MetricProducer producer)
+    {
+        metricProducers.add(producer);
+    }
+
+    public void addMetricProducer(final ComplexMetricProducer producer)
+    {
+        metricProducersComplex.add(producer);
+    }
+
+    public void addRawMetricProducer(final MetricProducer producer)
+    {
+        rawMetricProducers.add(producer);
+    }
+
+    public void addRawMetricProducer(final ComplexMetricProducer producer)
+    {
+        rawMetricProducersComplex.add(producer);
+    }
+
+    // only for MetricMonitorTest
+    void setSleepInterval(final long sleepInterval)
+    {
+        this.sleepIntervalMilliSeconds = sleepInterval;
+    }
+
+    public void start()
+    {
+        running.set(true);
+        monitorThread = new Thread(this::loop);
```

Comment:
Maybe it's easier to manage a thread with ExecutorService.

Reply:
It might be. But the current code is also not so complex, and it's easier to test.